lib/opte/src/engine/geneve.rs

```diff
@@ -245,17 +303,195 @@ impl From<&GeneveMeta> for GeneveHdrRaw {
+    /// given.
+    pub fn parse_all(
+        mut src: &[u8],
+        mut meta: Option<&mut GeneveMeta>,
+    ) -> Result<(), GeneveHdrError> {
+        while !src.is_empty() {
+            let option = GeneveOption::parse(&mut src)?;
+            if let Some(ref mut meta) = meta {
+                #[allow(clippy::single_match)]
+                match option {
+                    Some(GeneveOption::Oxide(OxideOption::External)) => {
+                        meta.oxide_external_pkt = true
+                    }
+                    _ => {}
+                }
+            }
+        }
+
+        Ok(())
+    }
+
+    /// Parse an individual Geneve option from a byte slice, advancing the
+    /// read location.
+    pub fn parse(src: &mut &[u8]) -> Result<Option<Self>, GeneveHdrError> {
+        let (head, tail) = src.split_at(GeneveOptHdrRaw::SIZE);
+        let opt_header = GeneveOptHdrRaw::new(head)?;
+        let needed_bytes = opt_header.options_len_bytes() as usize;
+        if tail.len() < needed_bytes {
+            return Err(GeneveHdrError::BadLength { len: needed_bytes as u16 });
+        }
+
+        let class = u16::from_be_bytes(opt_header.option_class);
+        let opt_type = opt_header.option_type();
+
+        // We don't yet have any options which need body parsing.
+        // This will skip over them regardless.
+        let (_body, tail) = tail.split_at(needed_bytes);
+        *src = tail;
+
+        // XXX: Break this out into a trait/impls to handle more cleanly.
+        Ok(match (class, opt_header.option_type()) {
+            (GENEVE_OPT_CLASS_OXIDE, 0) => {
+                Some(GeneveOption::Oxide(OxideOption::External))
+            }
+            _ if opt_header.is_critical() => {
+                return Err(GeneveHdrError::UnknownCriticalOption {
+                    class,
+                    opt_type,
+                })
+            }
+            _ => None,
+        })
+    }
+
+    /// Return the wire-length of this option in bytes, including headers.
+    pub fn len(&self) -> usize {
+        4 + match self {
+            GeneveOption::Oxide(o) => o.len(),
+        }
+    }
+
+    /// Emit an option, returning the number of 4-byte chunks written.
+    pub fn emit(&self, dst: &mut [u8]) -> usize {
+        let mut raw = GeneveOptHdrRaw::new_mut(dst).unwrap();
+
+        let (class, opt_type, len) = match self {
+            Self::Oxide(o) => (
+                GENEVE_OPT_CLASS_OXIDE,
+                o.opt_type(),
+                o.len() >> GENEVE_OPT_LEN_SCALE_SHIFT,
+            ),
+        };
+        raw.option_class = class.to_be_bytes();
+        raw.crit_type = opt_type;
+        raw.reserved_len = len as u8;
+
+        len + 1
+    }
+}
+
+/// Geneve options defined by Oxide, [`GENEVE_OPT_CLASS_OXIDE`].
+#[non_exhaustive]
+pub enum OxideOption {
+    /// A tag indicating that this packet originated from outside the VPC.
+    ///
+    /// Option Type `0`. Currently includes no body.
+    External,
+}
```

Comment:
Need to update https://github.com/oxidecomputer/oana#geneve-options with the new option type.

Reply:
Okay, I have a [PR opened](https://github.com/oxidecomputer/oana/pull/20).